Creating an audio unit in iOS simulator on a VM may trigger a TCC prompt if audio session is play-and-record

youennf opened the pull request:
#### 81be5526c807ef08fd41dd0084d6a1e68cb1ac46
<pre>
Creating an audio unit in iOS simulator on a VM may trigger a TCC prompt if audio session is play-and-record
<a href="https://bugs.webkit.org/show_bug.cgi?id=292944">https://bugs.webkit.org/show_bug.cgi?id=292944</a>
<a href="https://rdar.apple.com/132763275">rdar://132763275</a>

Reviewed by Eric Carlson.

When running iOS simulator in a VM, a TCC prompt for microphone happens in the following case:
- AudioSession is active
- AudioSession category is play-and-record
- An audio unit is created to render audio (for instance web audio).

To prevent the TCC prompt, we are now, in iOS simulator using playback category when we should use play-and-record.
We still report play-and-record in that case via a boolean stored in AudioSessionIOS.

* Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::setCategory):
(WebCore::AudioSessionIOS::category const):

Canonical link: <a href="https://commits.webkit.org/294961@main">https://commits.webkit.org/294961@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/246372431fa75dd3c62961c2bdb158d52827b140

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103625 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13628 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108801 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54272 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23663 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78746 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106631 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18358 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59081 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11529 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53626 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11588 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111179 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30772 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22695 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87740 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31133 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89682 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87388 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22245 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32263 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9986 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25127 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30700 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36012 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30500 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33831 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32061 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
```diff
@@ -81,6 +81,9 @@ class AudioSessionIOS final : public AudioSessionCocoa {
     String m_sceneIdentifier;
     SoundStageSize m_soundStageSize { SoundStageSize::Automatic };
+#if PLATFORM(IOS_FAMILY_SIMULATOR)
+    bool m_isPlayAndRecord { false };
```
Nit: It would be helpful to have the name make it clear this is an override. Maybe something like `m_dontUsePlayAndRecordForTesting`, or `m_fakePlayAndRecordForTesting`

youennf replied:
Indeed, I'll use `m_isFakingPlayAndRecordForTesting`

youennf committed:
Creating an audio unit in iOS simulator on a VM may trigger a TCC prompt if audio session is play-and-record
https://bugs.webkit.org/show_bug.cgi?id=292944
rdar://132763275

Reviewed by Eric Carlson.

When running iOS simulator in a VM, a TCC prompt for microphone happens in the following case:
- AudioSession is active
- AudioSession category is play-and-record
- An audio unit is created to render audio (for instance web audio).

To prevent the TCC prompt, we are now, in iOS simulator using playback category when we should use play-and-record.
We still report play-and-record in that case via a boolean stored in AudioSessionIOS.

* Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::setCategory):
(WebCore::AudioSessionIOS::category const):

Canonical link: https://commits.webkit.org/294961@main
